In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [2]:
data = pd.read_excel('salesworkload.xlsx')

In [3]:
df = data.copy()

In [4]:
df.head()

Unnamed: 0  Unnamed: 1      Unnamed: 2 Unnamed: 3  Unnamed: 4 Unnamed: 5  \
0  MonthYear  Time index         Country    StoreID        City    Dept_ID   
1    10.2016           1  United Kingdom      88253  London (I)          1   
2    10.2016           1  United Kingdom      88253  London (I)          2   
3    10.2016           1  United Kingdom      88253  London (I)          3   
4    10.2016           1  United Kingdom      88253  London (I)          4   

   Unnamed: 6 Unnamed: 7  Unnamed: 8   Unnamed: 9 Unnamed: 10 Unnamed: 11  \
0  Dept. Name   HoursOwn  HoursLease  Sales units    Turnover    Customer   
1         Dry   3184.764           0       398560     1226244         NaN   
2      Frozen   1582.941           0        82725      387810         NaN   
3       other     47.205           0       438400      654657         NaN   
4        Fish   1623.852           0       309425      499434         NaN   

  Unnamed: 12    Unnamed: 13  
0   Area (m2)  Opening hours  
1      953.04         Type A  
2      720.48         Type A  
3      966.72         Type A  
4     1053.36         Type A

In [5]:
df.tail()

Unnamed: 0 Unnamed: 1 Unnamed: 2 Unnamed: 3  Unnamed: 4 Unnamed: 5  \
7654    06.2017          9     Sweden      29650  Gothenburg         12   
7655    06.2017          9     Sweden      29650  Gothenburg         16   
7656    06.2017          9     Sweden      29650  Gothenburg         11   
7657    06.2017          9     Sweden      29650  Gothenburg         17   
7658    06.2017          9     Sweden      29650  Gothenburg         18   

             Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9 Unnamed: 10  \
7654           Checkout   6322.323          0    3886530    14538825   
7655  Customer Services   4270.479          0        245           0   
7656           Delivery          0          0          0           0   
7657             others   2224.929          0        245           0   
7658                all    39652.2          0    3886530    15056214   

     Unnamed: 11 Unnamed: 12 Unnamed: 13  
7654         NaN         #NV      Type A  
7655         NaN         #NV      Type A  
7656         NaN         #NV      Type A  
7657         NaN         #NV      Type A  
7658         NaN         #NV      Type A

In [6]:
df.shape

(7659, 14)

# Data Shape
    -7659 rows
        The first row has the names of the columns
    - 14 columns

In [7]:
#Rename the columns
columns = ['MonthYear', 'Time index', 'Country', 'StoreID', 'City', 'Dept_ID', 'Dept.Name', 'HoursOwn', 
           'HoursLease', 'Sales units', 'Turnover', 'Customer', 'Area (m2)', 'Opening hours']
df.columns = columns
display(df)

MonthYear  Time index         Country  StoreID        City  Dept_ID  \
0     MonthYear  Time index         Country  StoreID        City  Dept_ID   
1       10.2016           1  United Kingdom    88253  London (I)        1   
2       10.2016           1  United Kingdom    88253  London (I)        2   
3       10.2016           1  United Kingdom    88253  London (I)        3   
4       10.2016           1  United Kingdom    88253  London (I)        4   
...         ...         ...             ...      ...         ...      ...   
7654    06.2017           9          Sweden    29650  Gothenburg       12   
7655    06.2017           9          Sweden    29650  Gothenburg       16   
7656    06.2017           9          Sweden    29650  Gothenburg       11   
7657    06.2017           9          Sweden    29650  Gothenburg       17   
7658    06.2017           9          Sweden    29650  Gothenburg       18   

              Dept.Name  HoursOwn  HoursLease  Sales units  Turnover  \
0            Dept. Name  HoursOwn  HoursLease  Sales units  Turnover   
1                   Dry  3184.764           0       398560   1226244   
2                Frozen  1582.941           0        82725    387810   
3                 other    47.205           0       438400    654657   
4                  Fish  1623.852           0       309425    499434   
...                 ...       ...         ...          ...       ...   
7654           Checkout  6322.323           0      3886530  14538825   
7655  Customer Services  4270.479           0          245         0   
7656           Delivery         0           0            0         0   
7657             others  2224.929           0          245         0   
7658                all   39652.2           0      3886530  15056214   

      Customer  Area (m2)  Opening hours  
0     Customer  Area (m2)  Opening hours  
1          NaN     953.04         Type A  
2          NaN     720.48         Type A  
3          NaN     966.72         Type A  
4          NaN    1053.36         Type A  
...        ...        ...            ...  
7654       NaN        #NV         Type A  
7655       NaN        #NV         Type A  
7656       NaN        #NV         Type A  
7657       NaN        #NV         Type A  
7658       NaN        #NV         Type A  

[7659 rows x 14 columns]

In [8]:
#Drop the first row
df.drop(0, axis=0, inplace=True)
df

MonthYear Time index         Country StoreID        City Dept_ID  \
1      10.2016          1  United Kingdom   88253  London (I)       1   
2      10.2016          1  United Kingdom   88253  London (I)       2   
3      10.2016          1  United Kingdom   88253  London (I)       3   
4      10.2016          1  United Kingdom   88253  London (I)       4   
5      10.2016          1  United Kingdom   88253  London (I)       5   
...        ...        ...             ...     ...         ...     ...   
7654   06.2017          9          Sweden   29650  Gothenburg      12   
7655   06.2017          9          Sweden   29650  Gothenburg      16   
7656   06.2017          9          Sweden   29650  Gothenburg      11   
7657   06.2017          9          Sweden   29650  Gothenburg      17   
7658   06.2017          9          Sweden   29650  Gothenburg      18   

                Dept.Name  HoursOwn HoursLease Sales units  Turnover Customer  \
1                     Dry  3184.764          0      398560   1226244      NaN   
2                  Frozen  1582.941          0       82725    387810      NaN   
3                   other    47.205          0      438400    654657      NaN   
4                    Fish  1623.852          0      309425    499434      NaN   
5     Fruits & Vegetables  1759.173          0      165515    329397      NaN   
...                   ...       ...        ...         ...       ...      ...   
7654             Checkout  6322.323          0     3886530  14538825      NaN   
7655    Customer Services  4270.479          0         245         0      NaN   
7656             Delivery         0          0           0         0      NaN   
7657               others  2224.929          0         245         0      NaN   
7658                  all   39652.2          0     3886530  15056214      NaN   

     Area (m2) Opening hours  
1       953.04        Type A  
2       720.48        Type A  
3       966.72        Type A  
4      1053.36        Type A  
5      1053.36        Type A  
...        ...           ...  
7654       #NV        Type A  
7655       #NV        Type A  
7656       #NV        Type A  
7657       #NV        Type A  
7658       #NV        Type A  

[7658 rows x 14 columns]

In [9]:
df.shape

(7658, 14)

# Data Shape
    -7658 rows
        After deleting the first row
    -14 columns

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7658 entries, 1 to 7658
Data columns (total 14 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   MonthYear      7658 non-null   object
 1   Time index     7650 non-null   object
 2   Country        7650 non-null   object
 3   StoreID        7650 non-null   object
 4   City           7650 non-null   object
 5   Dept_ID        7650 non-null   object
 6   Dept.Name      7650 non-null   object
 7   HoursOwn       7650 non-null   object
 8   HoursLease     7650 non-null   object
 9   Sales units    7650 non-null   object
 10  Turnover       7650 non-null   object
 11  Customer       0 non-null      object
 12  Area (m2)      7650 non-null   object
 13  Opening hours  7650 non-null   object
dtypes: object(14)
memory usage: 837.7+ KB


In [11]:
#Checking for nulls
null = df.isnull().sum()
ratio = null / df.shape[0] * 100
pd.DataFrame({'Null_Sum': null, 'Ratio': ratio})

Null_Sum       Ratio
MonthYear             0    0.000000
Time index            8    0.104466
Country               8    0.104466
StoreID               8    0.104466
City                  8    0.104466
Dept_ID               8    0.104466
Dept.Name             8    0.104466
HoursOwn              8    0.104466
HoursLease            8    0.104466
Sales units           8    0.104466
Turnover              8    0.104466
Customer           7658  100.000000
Area (m2)             8    0.104466
Opening hours         8    0.104466

# Null Values
    -The 'Customer' column has 100% null values so it will be droped
    -The 'MonthYear' column has zero null values
    -The rest of the columns have 8 null values

In [12]:
#Drop the 'Customer' column
df.drop(columns=['Customer'], axis=1, inplace=True)
display(df)

MonthYear Time index         Country StoreID        City Dept_ID  \
1      10.2016          1  United Kingdom   88253  London (I)       1   
2      10.2016          1  United Kingdom   88253  London (I)       2   
3      10.2016          1  United Kingdom   88253  London (I)       3   
4      10.2016          1  United Kingdom   88253  London (I)       4   
5      10.2016          1  United Kingdom   88253  London (I)       5   
...        ...        ...             ...     ...         ...     ...   
7654   06.2017          9          Sweden   29650  Gothenburg      12   
7655   06.2017          9          Sweden   29650  Gothenburg      16   
7656   06.2017          9          Sweden   29650  Gothenburg      11   
7657   06.2017          9          Sweden   29650  Gothenburg      17   
7658   06.2017          9          Sweden   29650  Gothenburg      18   

                Dept.Name  HoursOwn HoursLease Sales units  Turnover  \
1                     Dry  3184.764          0      398560   1226244   
2                  Frozen  1582.941          0       82725    387810   
3                   other    47.205          0      438400    654657   
4                    Fish  1623.852          0      309425    499434   
5     Fruits & Vegetables  1759.173          0      165515    329397   
...                   ...       ...        ...         ...       ...   
7654             Checkout  6322.323          0     3886530  14538825   
7655    Customer Services  4270.479          0         245         0   
7656             Delivery         0          0           0         0   
7657               others  2224.929          0         245         0   
7658                  all   39652.2          0     3886530  15056214   

     Area (m2) Opening hours  
1       953.04        Type A  
2       720.48        Type A  
3       966.72        Type A  
4      1053.36        Type A  
5      1053.36        Type A  
...        ...           ...  
7654       #NV        Type A  
7655       #NV        Type A  
7656       #NV        Type A  
7657       #NV        Type A  
7658       #NV        Type A  

[7658 rows x 13 columns]

In [13]:
for i in df.columns:
    print(df[i].value_counts())
    print('*************************')

MonthYear
10.2016      850
11.2016      850
12.2016      850
01.2017      850
02.2017      850
03.2017      850
04.2017      850
05.2017      850
06.2017      850
 - - - -       8
Name: count, dtype: int64
*************************
Time index
1    850
2    850
3    850
4    850
5    850
6    850
7    850
8    850
9    850
Name: count, dtype: int64
*************************
Country
United Kingdom     918
Germany            918
France             918
Spain              765
Italy              765
Poland             612
The Netherlands    612
Czech Republic     612
Denmark            612
Belgium            459
Sweden             459
Name: count, dtype: int64
*************************
StoreID
88253    153
86089    153
78450    153
94153    153
64983    153
77348    153
78325    153
83160    153
12227    153
94882    153
34378    153
42367    153
98422    153
38976    153
79785    153
63354    153
85124    153
73422    153
91973    153
19340    153
76852    153
73762    153
81473    153
9099

##### The 'MonthYear' column has 8 of ' - - - - ' values that needs to be dropped
##### The 'Area (m2)' column has 850 of '#NV' values that will be droped

In [14]:
#drop the empty values in 'MonthYear' column
df = df[df['MonthYear'] != ' - - - - ']
df

MonthYear Time index         Country StoreID        City Dept_ID  \
1      10.2016          1  United Kingdom   88253  London (I)       1   
2      10.2016          1  United Kingdom   88253  London (I)       2   
3      10.2016          1  United Kingdom   88253  London (I)       3   
4      10.2016          1  United Kingdom   88253  London (I)       4   
5      10.2016          1  United Kingdom   88253  London (I)       5   
...        ...        ...             ...     ...         ...     ...   
7654   06.2017          9          Sweden   29650  Gothenburg      12   
7655   06.2017          9          Sweden   29650  Gothenburg      16   
7656   06.2017          9          Sweden   29650  Gothenburg      11   
7657   06.2017          9          Sweden   29650  Gothenburg      17   
7658   06.2017          9          Sweden   29650  Gothenburg      18   

                Dept.Name  HoursOwn HoursLease Sales units  Turnover  \
1                     Dry  3184.764          0      398560   1226244   
2                  Frozen  1582.941          0       82725    387810   
3                   other    47.205          0      438400    654657   
4                    Fish  1623.852          0      309425    499434   
5     Fruits & Vegetables  1759.173          0      165515    329397   
...                   ...       ...        ...         ...       ...   
7654             Checkout  6322.323          0     3886530  14538825   
7655    Customer Services  4270.479          0         245         0   
7656             Delivery         0          0           0         0   
7657               others  2224.929          0         245         0   
7658                  all   39652.2          0     3886530  15056214   

     Area (m2) Opening hours  
1       953.04        Type A  
2       720.48        Type A  
3       966.72        Type A  
4      1053.36        Type A  
5      1053.36        Type A  
...        ...           ...  
7654       #NV        Type A  
7655       #NV        Type A  
7656       #NV        Type A  
7657       #NV        Type A  
7658       #NV        Type A  

[7650 rows x 13 columns]

In [15]:
#checking the column values after droping the empty rows
df['MonthYear'].unique()

array(['10.2016', '11.2016', '12.2016', '01.2017', '02.2017', '03.2017',
       '04.2017', '05.2017', '06.2017'], dtype=object)

In [16]:
#splitting the 'MonthYear' column to 'Month' and 'Year' columns
df['Month'] = df['MonthYear'].apply(lambda x: int(x.split('.')[0]))
df['Year'] = df['MonthYear'].apply(lambda x: int(x.split('.')[1]))
df

MonthYear Time index         Country StoreID        City Dept_ID  \
1      10.2016          1  United Kingdom   88253  London (I)       1   
2      10.2016          1  United Kingdom   88253  London (I)       2   
3      10.2016          1  United Kingdom   88253  London (I)       3   
4      10.2016          1  United Kingdom   88253  London (I)       4   
5      10.2016          1  United Kingdom   88253  London (I)       5   
...        ...        ...             ...     ...         ...     ...   
7654   06.2017          9          Sweden   29650  Gothenburg      12   
7655   06.2017          9          Sweden   29650  Gothenburg      16   
7656   06.2017          9          Sweden   29650  Gothenburg      11   
7657   06.2017          9          Sweden   29650  Gothenburg      17   
7658   06.2017          9          Sweden   29650  Gothenburg      18   

                Dept.Name  HoursOwn HoursLease Sales units  Turnover  \
1                     Dry  3184.764          0      398560   1226244   
2                  Frozen  1582.941          0       82725    387810   
3                   other    47.205          0      438400    654657   
4                    Fish  1623.852          0      309425    499434   
5     Fruits & Vegetables  1759.173          0      165515    329397   
...                   ...       ...        ...         ...       ...   
7654             Checkout  6322.323          0     3886530  14538825   
7655    Customer Services  4270.479          0         245         0   
7656             Delivery         0          0           0         0   
7657               others  2224.929          0         245         0   
7658                  all   39652.2          0     3886530  15056214   

     Area (m2) Opening hours  Month  Year  
1       953.04        Type A     10  2016  
2       720.48        Type A     10  2016  
3       966.72        Type A     10  2016  
4      1053.36        Type A     10  2016  
5      1053.36        Type A     10  2016  
...        ...           ...    ...   ...  
7654       #NV        Type A      6  2017  
7655       #NV        Type A      6  2017  
7656       #NV        Type A      6  2017  
7657       #NV        Type A      6  2017  
7658       #NV        Type A      6  2017  

[7650 rows x 15 columns]

In [17]:
#drop the 'MonthYear' column
df.drop(columns=['MonthYear'], inplace=True)
df

Time index         Country StoreID        City Dept_ID  \
1             1  United Kingdom   88253  London (I)       1   
2             1  United Kingdom   88253  London (I)       2   
3             1  United Kingdom   88253  London (I)       3   
4             1  United Kingdom   88253  London (I)       4   
5             1  United Kingdom   88253  London (I)       5   
...         ...             ...     ...         ...     ...   
7654          9          Sweden   29650  Gothenburg      12   
7655          9          Sweden   29650  Gothenburg      16   
7656          9          Sweden   29650  Gothenburg      11   
7657          9          Sweden   29650  Gothenburg      17   
7658          9          Sweden   29650  Gothenburg      18   

                Dept.Name  HoursOwn HoursLease Sales units  Turnover  \
1                     Dry  3184.764          0      398560   1226244   
2                  Frozen  1582.941          0       82725    387810   
3                   other    47.205          0      438400    654657   
4                    Fish  1623.852          0      309425    499434   
5     Fruits & Vegetables  1759.173          0      165515    329397   
...                   ...       ...        ...         ...       ...   
7654             Checkout  6322.323          0     3886530  14538825   
7655    Customer Services  4270.479          0         245         0   
7656             Delivery         0          0           0         0   
7657               others  2224.929          0         245         0   
7658                  all   39652.2          0     3886530  15056214   

     Area (m2) Opening hours  Month  Year  
1       953.04        Type A     10  2016  
2       720.48        Type A     10  2016  
3       966.72        Type A     10  2016  
4      1053.36        Type A     10  2016  
5      1053.36        Type A     10  2016  
...        ...           ...    ...   ...  
7654       #NV        Type A      6  2017  
7655       #NV        Type A      6  2017  
7656       #NV        Type A      6  2017  
7657       #NV        Type A      6  2017  
7658       #NV        Type A      6  2017  

[7650 rows x 14 columns]

In [18]:
#drop the '#NV' values in 'Area (m2)' column
df = df[df['Area (m2)'] != '#NV']
df

Time index         Country StoreID        City Dept_ID  \
1             1  United Kingdom   88253  London (I)       1   
2             1  United Kingdom   88253  London (I)       2   
3             1  United Kingdom   88253  London (I)       3   
4             1  United Kingdom   88253  London (I)       4   
5             1  United Kingdom   88253  London (I)       5   
...         ...             ...     ...         ...     ...   
6803          8          Sweden   29650  Gothenburg      12   
6804          8          Sweden   29650  Gothenburg      16   
6805          8          Sweden   29650  Gothenburg      11   
6806          8          Sweden   29650  Gothenburg      17   
6807          8          Sweden   29650  Gothenburg      18   

                Dept.Name   HoursOwn HoursLease Sales units  Turnover  \
1                     Dry   3184.764          0      398560   1226244   
2                  Frozen   1582.941          0       82725    387810   
3                   other     47.205          0      438400    654657   
4                    Fish   1623.852          0      309425    499434   
5     Fruits & Vegetables   1759.173          0      165515    329397   
...                   ...        ...        ...         ...       ...   
6803             Checkout   6668.493          0     3793520  14025312   
6804    Customer Services   4506.504          0         245         0   
6805             Delivery          0          0           0         0   
6806               others   2640.333          0         245         0   
6807                  all  45776.262          0     3793520  13823949   

     Area (m2) Opening hours  Month  Year  
1       953.04        Type A     10  2016  
2       720.48        Type A     10  2016  
3       966.72        Type A     10  2016  
4      1053.36        Type A     10  2016  
5      1053.36        Type A     10  2016  
...        ...           ...    ...   ...  
6803  31308.96        Type A      5  2017  
6804         0        Type A      5  2017  
6805         0        Type A      5  2017  
6806         0        Type A      5  2017  
6807  31308.96        Type A      5  2017  

[6800 rows x 14 columns]

##### from this table each department has name and id that represents it, i think the department names can be removed

In [19]:
dept = pd.DataFrame([list(df['Dept.Name'].unique()), list(df['Dept_ID'].unique())]).T
dept.columns = ['Dept_Name', 'Dept_ID']
dept

Dept_Name Dept_ID
0                   Dry       1
1                Frozen       2
2                 other       3
3                  Fish       4
4   Fruits & Vegetables       5
5                  Meat       6
6                  Food      13
7              Clothing       7
8             Household       8
9              Hardware       9
10             Non Food      14
11                Admin      15
12             Checkout      12
13    Customer Services      16
14             Delivery      11
15               others      17
16                  all      18

##### i think the city has duplicates for some cities like London (I) and London (II) or they may actually represent different things but because i have no context for the data i can't say how i will deal with it in this case

In [20]:
df['City'].unique()

array(['London (I)', 'Manchester', 'Liverpool', 'Birmingham', 'Leicester',
       'London (II)', 'Warsaw (I)', 'Warsaw (II)', 'Poznan', 'Krakow',
       'Amsterdam', 'Den Haag', 'Rotterdam', 'Groningen', 'Prague (I)',
       'Brno', 'Ostrava', 'Prague (II)', 'Copenhagen (I)',
       'Copenhagen (II)', 'Aalborg (I)', 'Aalborg (II)', 'Madrid (I)',
       'Madrid (II)', 'Barcelona (I)', 'Barcelona (II)', 'Bilbao',
       'Rome (I)', 'Rome (II)', 'Milano', 'Bologna', 'Napoli',
       'Berlin (I)', 'Berlin (II)', 'Munich', 'Hamburg', 'Frankfurt',
       'Cologne', 'Paris (I)', 'Paris (II)', 'Marseille', 'Lyon',
       'Bordeaux', 'Nantes', 'Brussels (I)', 'Brussels (II)', 'Antwerp',
       'Stockholm', 'Malmö', 'Gothenburg'], dtype=object)

In [21]:
#there was an error while converting the data type of 'HoursOwn' column, had to drop the rows with the '?' value in it
df = df[df['HoursOwn'] != '?']

In [22]:
int_cols = ['StoreID', 'Dept_ID', 'HoursLease', 'Sales units', 'Turnover']
float_cols = ['HoursOwn', 'Area (m2)']

df[int_cols] = df[int_cols].astype('int')
df[float_cols] = df[float_cols].astype('float')

df.dtypes

Time index        object
Country           object
StoreID            int32
City              object
Dept_ID            int32
Dept.Name         object
HoursOwn         float64
HoursLease         int32
Sales units        int32
Turnover           int32
Area (m2)        float64
Opening hours     object
Month              int64
Year               int64
dtype: object

### Data Summary
    -'Time index' (numerical): the data has 9 different indices [1-9] 
    -'Country' (categorical): country where the store is located
    -'StoreID' (numerical): id for each store
    -'City' (categorical): city where the store is located
    -'Dept_ID' (numerical): id for each department in the store
    -'Dept.Name' (categorical): name for each department in the store
    - HoursOwn (numerical): total number of operating hours for the store’s owned departments
    -'HoursLease' (numerical): total number of operating hours for leased departments
    -'Sales units' (numerical): number of units sold in each department 
    -'Turnover' (numerical): total earnings in each department
    -'Area (m2)' (numerical): area in square meters
    -'Opening hours' (categorical): type of store opening ['Type A', 'Type B', 'Type C'] 
    -'Month' (numerical): month or recorded data
    -'Year' (numerical): year of recorded data